### ISSUE TITLE
v4.50.3 Update Fails

### ISSUE THREAD

**Issue by petebocken** [homebridge/homebridge-config-ui-x#1493]
### Describe The Bug

Clicking the update button on the web for 4.50.3 and it is failing.

### Logs

```text
CMD: npm install -g homebridge-config-ui-x@4.50.3

npm ERR! process terminated
npm ERR! signal SIGTERM

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/homebridge/.npm/_logs/2023-05-12T13_35_33_771Z-debug-0.log

Cleaning up npm cache, please wait...
npm cache cleared, please try updating homebridge-config-ui-x again.

Operation failed. Please review log for details.


2110 verbose stack Error: process terminated
2110 verbose stack     at /opt/homebridge/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/reify.js:200:39
2110 verbose stack     at process.<anonymous> (/opt/homebridge/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/signal-handling.js:53:5)
2110 verbose stack     at process.emit (node:events:525:35)
```

### Homebridge UI Version

4.50.2

### Homebridge Version

1.6.1

### Node.js Version

16.14.0

### Operating System

Raspberry Pi OS / Raspbian

### Environment Info

- [ ] Using Docker?
- [ ] Using Hyper-V?
- [ ] Using hb-service?

### Raspberry Pi Model

Raspberry Pi Zero W

**Comment by hlong0712** [comment]
Yes , i click update and can not access homebridge over local web .

**Comment by petebocken** [comment]
> Yes , i click update and can not access homebridge over local web .

My homebridge install is still functional. The update just fails to complete.

**Comment by jca2112** [comment]
This homebridge-config-ui-x update also took down my Homebridge instance. Currently trying to uninstall/reinstall the homebridge-config-ui-x, but has not worked so far.

**Comment by hlong0712** [comment]
> This homebridge-config-ui-x update also took down my Homebridge instance. Currently trying to uninstall/reinstall the homebridge-config-ui-x, but has not worked so far.

Yes , i need uninstall and reinstall from fresh

**Comment by jca2112** [comment]
> Yes , i need uninstall and reinstall from fresh

@hlong0712 what uninstall & install command(s) (in which OS) did you use? Was it just for _homebridge-config-ui-x_?

**Comment by NorthernMan54** [comment]
@petebocken the working theory for why the install failed for you was that your RPI may have been low of memory needed to run the install process.  How much free ram does the dashboard show before you attempt the upgrade.

**Comment by petebocken** [comment]
> @petebocken the working theory for why the install failed for you was that your RPI may have been low of memory needed to run the install process. How much free ram does the dashboard show before you attempt the upgrade.

0.42 GB Total
0.23 GB Available

I did already try restarting the Pi and performing the update again.

**Comment by NorthernMan54** [comment]
@petebocken just to confirm that the issue is environmental and not the release packaging could you install the previous version?

You can select alternate versions to install with the wrench icon on the homebridge-config-ui-x plugin widget

**Comment by Staubgeborener** [comment]
@NorthernMan54 so you are the new main contributor for this project? Do you also plan to update the docker version?

**Comment by NorthernMan54** [comment]
@Staubgeborener All the homebridge collaborators are now contributing to the ongoing maintenance and support of this package and several other core packages, including the docker image.

I just happen to have the been person to hit the release button at this time, so I’m also taking the lead on post release support.

Once this release has settled into the community the plan is to also update all the packages like the docker include that include this.  As I wanted to reduce the release risk, I’m staggering the packages.

**Comment by petebocken** [comment]
> @petebocken just to confirm that the issue is environmental and not the release packaging could you install the previous version?

Downgraded from 4.50.2 to 4.50.1 just fine.
Upgraded from 4.50.1 to 4.50.2 just fine.
Upgrade from 4.50.2 to 4.50.3 fails with same error.

**Comment by NorthernMan54** [comment]
@petebocken Tks let me take a deeper dive into this.

**Comment by bwp91** [comment]
If you use the homebridge terminal to try to update to the new version of the ui

```
npm install -g homebridge-config-ui-x@4.50.3
```

does this complete or result in the same error: ?

```
npm ERR! process terminated
npm ERR! signal SIGTERM
```

**Comment by petebocken** [comment]
> If you use the homebridge terminal to try to update to the new version of the ui
> 
> ```
> npm install -g homebridge-config-ui-x@4.50.3
> ```
> 
> does this complete or result in the same error: ?

Here's what I got. I see it needs a newer version of node. I tried updating it, but not sure I'm doing it correctly. Can't get it to update. I got npm updated to 9.6.6.
```
$ sudo npm install -g homebridge-config-ui-x@4.50.3
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'homebridge-config-ui-x@4.50.3',
npm WARN EBADENGINE   required: { node: '>=16.20.0', homebridge: '>=1.6.0' },
npm WARN EBADENGINE   current: { node: 'v16.14.0', npm: '8.6.0' }
npm WARN EBADENGINE }
npm notice
npm notice New major version of npm available! 8.6.0 -> 9.6.6
npm notice Changelog: https://github.com/npm/cli/releases/tag/v9.6.6
npm notice Run npm install -g npm@9.6.6 to update!
npm notice
npm ERR! code ENOTEMPTY
npm ERR! syscall rename
npm ERR! path /usr/local/lib/node_modules/homebridge-config-ui-x
npm ERR! dest /usr/local/lib/node_modules/.homebridge-config-ui-x-3gtnOIj4
npm ERR! errno -39
npm ERR! ENOTEMPTY: directory not empty, rename '/usr/local/lib/node_modules/homebridge-config-ui-x' -> '/usr/local/lib/node_modules/.homebridge-config-ui-x-3gtnOIj4'

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2023-05-12T21_54_47_961Z-debug-0.log
```

**Comment by bwp91** [comment]
In this case I would try this:

```
rm -rf /usr/local/lib/node_modules/homebridge-config-ui-x
```

(if you get permission issues then maybe try with `sudo` at the beginning)

and then (without restarting homebridge in between) install the plugin

```
npm install -g homebridge-config-ui-x
```

**Comment by bwp91** [comment]
Your version of node (although you are getting warnings) I don't think is important for the install of the ui plugin.

Once you have config-ui-x running again on the latest version, you could consider using the `hb-service` tool to update your node to the latest 18.x version

**Comment by jca2112** [comment]
I've tried deleting the directory and reinstalling (as mentioned previously), but I am still unable to get the UI to load.

**Comment by bwp91** [comment]
sorry i didn’t see you mentioned earlier about deleting the directory

**Comment by bwp91** [comment]
i feel like maybe different users may have slightly different things going on and/or issues why the ui isn’t updating properly

**Comment by petebocken** [comment]
@bwp91 deleting the dir and reinstalling appears to have completed fine?  However, the web still shows I'm on 4.50.2.

```
$ sudo rm -rf /usr/local/lib/node_modules/homebridge-config-ui-x
pi@web:/usr/local/lib/node_modules $ sudo npm install -g homebridge-config-ui-x
/usr/local/bin/homebridge-config-ui-x -> /usr/local/lib/node_modules/homebridge-config-ui-x/dist/bin/standalone.js
/usr/local/bin/hb-service -> /usr/local/lib/node_modules/homebridge-config-ui-x/dist/bin/hb-service.js

> node-pty-prebuilt-multiarch@0.10.1-pre.5 install /usr/local/lib/node_modules/homebridge-config-ui-x/node_modules/node-pty-prebuilt-multiarch
> node scripts/check-prebuild.js || prebuild-install || node scripts/install.js

Prebuild binary exists: /usr/local/lib/node_modules/homebridge-config-ui-x/node_modules/node-pty-prebuilt-multiarch/prebuilds/linux-arm/node.abi67.node

> core-js-pure@3.30.2 postinstall /usr/local/lib/node_modules/homebridge-config-ui-x/node_modules/core-js-pure
> node -e "try{require('./postinstall')}catch(e){}"

> @nestjs/core@9.0.8 postinstall /usr/local/lib/node_modules/homebridge-config-ui-x/node_modules/@nestjs/core
> opencollective || exit 0

> node-pty-prebuilt-multiarch@0.10.1-pre.5 postinstall /usr/local/lib/node_modules/homebridge-config-ui-x/node_modules/node-pty-prebuilt-multiarch
> node scripts/post-install.js

npm WARN @nestjs/common@9.0.8 requires a peer of cache-manager@* but none is installed. You must install peer dependencies yourself.
npm WARN @nestjs/platform-fastify@9.0.8 requires a peer of @fastify/view@^7.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @nestjs/core@9.0.8 requires a peer of @nestjs/microservices@^9.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @nestjs/core@9.0.8 requires a peer of @nestjs/platform-express@^9.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN node-fetch@2.6.11 requires a peer of encoding@^0.1.0 but none is installed. You must install peer dependencies yourself.
npm WARN ws@8.2.3 requires a peer of bufferutil@^4.0.1 but none is installed. You must install peer dependencies yourself.
npm WARN ws@8.2.3 requires a peer of utf-8-validate@^5.0.2 but none is installed. You must install peer dependencies yourself.

+ homebridge-config-ui-x@4.50.3
added 346 packages from 398 contributors in 377.732s
```

**Comment by NorthernMan54** [comment]
@petebocken Could you try one more thing for us, we packaged the release again and published it as version 4.50.4.   We found the issue with the packaging of 4.50.3, and believe this will resolve the issue.

PS You may want to remove this version - `sudo rm -rf /usr/local/lib/node_modules/homebridge-config-ui-x`

**Comment by NorthernMan54** [comment]
For people watching this, low powered/low resource devices, like a api w or a rpi 2 use an alternate method of plugin installation that requires the GitHub release package to include a artifact bundle that includes the distributed package, and that step was missed with 4.50.3.

**Comment by petebocken** [comment]
Updating to 4.50.4 worked. 👍 thanks!

**Comment by jca2112** [comment]
Just wanted to follow-up in case this is still an issue for anyone else. I tried deleting and downgrading the UI version -- did not work. I then tried deleting and installing the just released 4.50.4 version, still did not work. I finally thought to reinstall the homebridge user, and that got the UI working:

`sudo hb-service install --user homebridge`

Now I'm having to reinstall plugins.

**Comment by NorthernMan54** [comment]
@jca2112 I think your issue may have been different than the one reported by @petebocken, in his case the install of version homebridge-config-ui-x 4.50.3 did not succeed and he was able to recover manually.

In your case I think the install corrupted homebridge-config-ui-x, and one method of recovery would be to try these steps.

```
sudo hb-shell
npm install -g homebridge-config-ui-x
```

The homebridge apt package creates its own location for the installation of homebridge-config-ui-x in /opt/homebridge/lib/node_modules/homebridge-config-ui-x.  And by using the hb-shell it set's the environment settings needed.
